docs/source/tutorials/InvertLensEquation.ipynb

```diff
@@ -219,10 +276,290 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "Ninit = 100\n",
+    "x_init = (torch.rand(Ninit) - 0.5) * fov\n",
+    "y_init = (torch.rand(Ninit) - 0.5) * fov\n",
+    "\n",
+    "\n",
+    "def raytrace(x, y):\n",
+    "    return lens.raytrace(x, y, z_s)\n",
+    "\n",
+    "\n",
+    "final = caustics.lenses.func.forward_raytrace_rootfind(\n",
+    "    x_init, y_init, sp_x, sp_y, raytrace\n",
+    ")\n",
+    "x_final, y_final = final[..., 0], final[..., 1]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2abb217e",
+   "metadata": {},
+   "source": [
+    "Here we easily find the four magnified images, but the central demagnified image is (often) not found by this method since a point has to get lucky enough to start very close to the correct position in order for the gradient based root finder to work."
+   ]
+  },
   {
    "cell_type": "code",
    "execution_count": null,
    "id": "d2fe0e6f",
+   "metadata": {
+    "tags": [
+     "hide-input"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "fig, ax = plt.subplots()\n",
+    "CS = ax.contour(thx, thy, detA, levels=[0.0], colors=\"b\", zorder=1)\n",
+    "for path in paths:\n",
+    "    # Collect the path into a discrete set of points\n",
+    "    x1 = torch.tensor(list(float(vs[0]) for vs in path))\n",
+    "    x2 = torch.tensor(list(float(vs[1]) for vs in path))\n",
+    "    # raytrace the points to the source plane\n",
+    "    y1, y2 = lens.raytrace(x1, x2, z_s)\n",
+    "\n",
+    "    # Plot the caustic\n",
+    "    ax.plot(y1, y2, color=\"r\", zorder=1)\n",
+    "colors = [\"tab:red\", \"tab:blue\", \"tab:green\", \"tab:orange\", \"tab:purple\"]\n",
+    "for c in colors:\n",
+    "    if x_final.shape[0] == 0:\n",
+    "        break\n",
+    "    R = ((x_final[0] - x_final) ** 2 + (y_final[0] - y_final) ** 2).sqrt()\n",
+    "    ax.scatter(x_init[R < 0.1], y_init[R < 0.1], color=c)\n",
+    "    ax.scatter(x_final[0], y_final[0], color=\"k\", s=200, marker=\"*\")\n",
+    "    ax.scatter(x_final[0], y_final[0], color=c, s=100, marker=\"*\")\n",
+    "    x_init = x_init[R >= 0.1]\n",
+    "    y_init = y_init[R >= 0.1]\n",
+    "    x_final = x_final[R >= 0.1]\n",
+    "    y_final = y_final[R >= 0.1]\n",
+    "ax.axes.set_axis_off()\n",
+    "ax.set_xlim([-fov / 1.9, fov / 1.9])\n",
+    "ax.set_ylim([-fov / 1.9, fov / 1.9])\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b4c5d47b",
+   "metadata": {},
+   "source": [
+    "Let's now look at a more clever algorithm. We will map triangles in the image plane to triangles in the source plane, we may then explore recursively, any triangles which enclose the desired source point. Due to the non-linearity of the gravitational lensing transformation, we will also search the neighbor of any triangle that seems to have found an image position. First we highlight in green, any triangles which contain the source point, then expand to all their neighbors."
```

Comment:
In the 2nd iteration, it looks like there are many more triangles included than just those that enclose the source (green triangles) and their nearest neighbors. Is that right? If so, I don't see why one would want to search beyond the immediate neighbors.
<img width="349" alt="image" src="https://github.com/user-attachments/assets/4e3da311-4e39-4753-97a9-3450db7f73b6">

Reply:
Ah so the neighbors are created with the larger triangles before upsampling them, so you would want to look at the figure before this one to see where the neighbors should be. Either way, yes this is searching a pretty large area, but the lens images can be quite elusive if we are unlucky, so this is just being cautious.